### ISSUE TITLE
Refactoring to convert to static member and vice versa

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#43917]
There is no 'make static' refactoring for methods, so I have to fix up the call sites by hand if I ever want to do that.

It also seems like there could be a code fix for "CS0176 Member cannot be accessed with an instance reference; qualify it with a type name instead."

@mavasani Cyrus said roslyn-analyzers has this and that I should ask you to port it.

**Comment by alrz** [comment]
Current make static is not that interesting. I think it only works if you missed `static` on a method in a static class?

what I find very helpful is to suggest to make static if there's no `this` reference. 

I think there exists such feature in roslyn-analyzers. Not sure if you're referring to that one or something else.

**Comment by mavasani** [comment]
Yes, as @alrz mentioned, [FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) implements [CA1822](https://docs.microsoft.com/visualstudio/code-quality/ca1822) to flag members that do not reference any instance members, but are not marked static. It also provides a code fix that makes the member static and also fixes up all call sites.

Our divide between whether something belongs to roslyn repo versus roslyn-analyzers repo is as follows:
- Roslyn repo:
   - Code style analyzers
   - Code fixes for compiler diagnostics
   - Pure code refactorings.

   This would be VS-built in functionality. Code style analyzers are also available via [CodeStyle NuGet package](https://dotnet.myget.org/feed/roslyn/package/nuget/Microsoft.CodeAnalysis.CSharp.CodeStyle) for build enforcement.
- Roslyn Analyzers repo:
   - Code quality analyzers performing semantic analysis, that are extremely likely to be enforced in build.

   Available as [CA rules](https://docs.microsoft.com/visualstudio/code-quality/code-analysis-warnings-for-managed-code-by-checkid) in [FxCopAnalyzers NuGet package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers).

So, if your request is for a code quality enforcement for marking members as static, whenever they could be, CA1822 already exists.
If your request is about any refactorings/code fixes that is just a helpful tool to mark something static without any semantic analysis, this should be written as a new refactoring in Roslyn repo.

**Comment by jnm2** [comment]
I'm not interested in enforcement or warnings, strictly refactoring help.

**Comment by mavasani** [comment]
> I'm not interested in enforcement or warnings, strictly refactoring help.

Thanks @jnm2. We should be able to port [MarkMembersAsStatic.Fixer.cs](https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/MarkMembersAsStatic.Fixer.cs) as a code refactoring in Roslyn repo. The core design/PR question is would this be too noisy to show this refactoring on every instance member?

**Comment by mavasani** [comment]
Marked this for next design review meeting so we can iron out when this refactoring should be shown.

**Comment by alrz** [comment]
>would it be too noisy to show this refactoring on every instance member?

I don't think so. Rider has set this to "suggestion" by default and most of the time static was preferable for me. that way I can know if I want to move the method to another type or make local function, since it doesn't depend on the instance.

**Comment by alrz** [comment]
>every instance member?

it wouldn't be on *every* instance member thought, only those that doesn't have an implicit or explicit reference to `this`.

**Comment by mavasani** [comment]
@alrz - what you are requesting is already available as CA1822 in FxCopAnalyzers package - you can easily configure the severity for CA1822 as a Suggestion/Info and get the exact same behavior. 
Duplicating the exact same functionality in Roslyn does not provide much benefit. Additionally, with .NET5 the CA analyzers will soon be shipping with the SDK (majority either disabled or enabled as suggestions). CA1822 will ship as a suggestion as part of this, so you will get it by default for all .NET5 projects and can enable CA rules in prior projects via some MSBuild property (still being designed).

To summarize, if the intent here is to see a suggestion/refactoring for Make Static only for instance members that can be safely made static, you can already do so currently by installing FxCopAnalyzers package, and soon this analyzer suggestion will be available with .NET SDK itself.
However, if the intent is to get a refactoring for cases where you are working towards converting an instance member into static member, where this instance member does reference other instance member and would need follow-up fixups to delete or refactor code that touches instance references to make sure it compiles, then this can be added as a new Roslyn refactoring.

**Comment by alrz** [comment]
>you can already do so currently by installing FxCopAnalyzers package

For instance, when I work on roslyn, I don't really want to install packages to get a refactoring suggestion. I just want it to be available in the IDE - that shouldn't depend on the project settings.

Right now, it seems like VS is lacking this feature, while it is bound to some code style package. IMO we shouldn't offer a (lightweight and generally useful) productivity feature only as a nuget package, rather it should be bundled within the IDE for the maximum accessibility.

**Comment by mavasani** [comment]
> when I work on roslyn

We should be able to enable CA1822 as a suggestion in Roslyn.sln. We just need to change the below to "Info": https://github.com/dotnet/roslyn/blob/d7d0c22e12f7684adf75b92e205b6b5bb7391a1d/eng/config/rulesets/Shipping.ruleset#L54

> Right now, it seems like VS is lacking this feature, while it is bound to some code style package. IMO we shouldn't offer a (lightweight and generally useful) productivity feature only as a nuget package, rather it should be bundled within the IDE for the maximum accessibility.

This is potentially the core feature request that needs to be discussed at the design meeting.

**Comment by mavasani** [comment]
> Right now, it seems like VS is lacking this feature, while it is bound to some code style package. IMO we shouldn't offer a (lightweight and generally useful) productivity feature only as a nuget package, rather it should be bundled within the IDE for the maximum accessibility.

Tagging @tmat as well. Given his recent refactoring in Roslyn to treat VS/host analyzer references as solution references that can execute in OOP, I think we can add a feature where analyzer NuGet packages are installed to a VS instance rather then a specific project to act similar to VSIX packages.

**Comment by JoeRobich** [comment]
## Design Review
- This is not really a code-style analyzer. If it were to move it would be as a refactoring.
- Any project can reference the Roslyn-Analyzers today, but there is friction if it isn't enabled for a project.
- Could package and ship this as a VSIX.

Proposal:
- Keep the analyzer in Roslyn-Analyzers, but enable it in the Roslyn repo.
- Consider enhancing the Change Signature dialog with the ability to make members static.

**Comment by mavasani** [comment]
> Keep the analyzer in Roslyn-Analyzers, but enable it in the Roslyn repo.

FYI: This part is complete. CA1822 is available as an IDE suggestion in entire Roslyn repo and enforced as a build warning in certain IDE layers (Workspaces, Features and CodeStyle).

**Comment by alrz** [comment]
>Could package and ship this as a VSIX.

Is this part going to happen? I think that is exactly what I've been asking for in any project out there that is not roslyn.
